File: render_frame.py
#!/usr/bin/env python3
import sys, os, json, textwrap
from PIL import Image, ImageDraw, ImageFont

def render_frame(persona_path, wisdom_text, persona_name, bar_heights, output_path, frame_type='persona'):
    W, H = 1080, 1920
    img = Image.new('RGB', (W, H), '#08080f')
    draw = ImageDraw.Draw(img)

    # Background image (persona or mudrets portrait)
    try:
        portrait = Image.open(persona_path).convert('RGB')
        portrait = portrait.resize((W, W), Image.NEAREST)
        img.paste(portrait, (0, 0))
        # dark overlay on portrait bottom
        overlay = Image.new('RGBA', (W, W), (8, 8, 15, 160))
        img.paste(Image.new('RGB', (W, W), '#08080f'), (0, 0), overlay)
    except Exception as e:
        print(f'Portrait error: {e}', file=sys.stderr)

    # Scanlines
    for y in range(0, W, 4):
        draw.line([(0, y), (W, y)], fill=(0, 0, 0, 25), width=1)

    # Gold border
    draw.rectangle([4, 4, W-4, H-4], outline='#c8a84b', width=6)

    # Persona name
    try:
        font_name = ImageFont.truetype('/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf', 42)
        font_text = ImageFont.truetype('/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf', 52)
        font_small = ImageFont.truetype('/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf', 32)
    except:
        font_name = ImageFont.load_default()
        font_text = font_name
        font_small = font_name

    name_y = W + 50
    draw.text((W//2, name_y), f'— {persona_name.upper()} —', font=font_name, fill='#c8a84b', anchor='mm')

    # Equalizer bars
    eq_y = W + 120
    bars = bar_heights if bar_heights else [8]*20
    n = len(bars)
    bar_w = 28
    gap = 14
    total = n * (bar_w + gap) - gap
    sx = (W - total) // 2
    for i, h in enumerate(bars):
        x = sx + i * (bar_w + gap)
        draw.rectangle([x, eq_y + 60 - h, x + bar_w, eq_y + 60], fill='#c8a84b')

    # Wisdom text
    text_y = eq_y + 100
    wrapped = textwrap.fill(wisdom_text, width=28)
    lines = wrapped.split('\n')
    for i, line in enumerate(lines):
        draw.text((W//2, text_y + i * 70), line, font=font_text, fill='#e8c878', anchor='mm')

    # Footer
    draw.text((W//2, H - 70), 'mudrets.online', font=font_small, fill='#3a3020', anchor='mm')

    img.save(output_path, 'PNG')

if __name__ == '__main__':
    data = json.loads(sys.argv[1])
    render_frame(
        data['persona_path'],
        data['wisdom_text'],
        data['persona_name'],
        data.get('bar_heights', [8]*20),
        data['output_path'],
        data.get('frame_type', 'persona')
    )

File: render_frame.js
// render_frame.js — рендерит один кадр 1080x1920 для TikTok видео
const { createCanvas, loadImage } = require('@napi-rs/canvas');
const fs = require('fs');

async function renderFrame(data) {
  const W = 1080, H = 1920;
  const canvas = createCanvas(W, H);
  const ctx = canvas.getContext('2d');

  // Background
  ctx.fillStyle = '#08080f';
  ctx.fillRect(0, 0, W, H);

  // Portrait
  try {
    const portrait = await loadImage(data.persona_path);
    ctx.drawImage(portrait, 0, 0, W, W);
    // dark gradient overlay bottom of portrait
    const grad = ctx.createLinearGradient(0, W * 0.6, 0, W);
    grad.addColorStop(0, 'rgba(8,8,15,0)');
    grad.addColorStop(1, 'rgba(8,8,15,0.85)');
    ctx.fillStyle = grad;
    ctx.fillRect(0, 0, W, W);
  } catch(e) {
    console.error('portrait error:', e.message);
  }

  // Scanlines
  ctx.fillStyle = 'rgba(0,0,0,0.07)';
  for (let y = 0; y < W; y += 4) ctx.fillRect(0, y, W, 1);

  // Gold border
  ctx.strokeStyle = '#c8a84b';
  ctx.lineWidth = 10;
  ctx.strokeRect(5, 5, W - 10, H - 10);

  // Persona name
  ctx.fillStyle = '#c8a84b';
  ctx.font = 'bold 52px monospace';
  ctx.textAlign = 'center';
  ctx.fillText(`— ${(data.persona_name || 'МУДРЕЦ').toUpperCase()} —`, W / 2, W + 80);

  // Equalizer bars
  const bars = data.bar_heights || Array(20).fill(8);
  const barW = 32, gap = 16;
  const totalBarW = bars.length * (barW + gap) - gap;
  const barStartX = (W - totalBarW) / 2;
  const barBaseY = W + 160;
  bars.forEach((h, i) => {
    const x = barStartX + i * (barW + gap);
    ctx.fillStyle = '#c8a84b';
    ctx.fillRect(x, barBaseY - h, barW, h);
  });

  // Wisdom text
  const text = data.wisdom_text || '';
  ctx.fillStyle = '#e8c878';
  ctx.font = '54px serif';
  ctx.textAlign = 'center';

  // Word wrap
  const maxWidth = W - 120;
  const lineHeight = 72;
  const words = text.split(' ');
  const lines = [];
  let line = '';
  for (const word of words) {
    const test = line + (line ? ' ' : '') + word;
    if (ctx.measureText(test).width > maxWidth && line) {
      lines.push(line);
      line = word;
    } else {
      line = test;
    }
  }
  if (line) lines.push(line);

  const textStartY = W + 220;
  lines.forEach((l, i) => ctx.fillText(l, W / 2, textStartY + i * lineHeight));

  // Footer
  ctx.fillStyle = '#2a2010';
  ctx.font = '36px monospace';
  ctx.fillText('mudrets.online', W / 2, H - 60);

  // Save
  const buf = canvas.toBuffer('image/png');
  fs.writeFileSync(data.output_path, buf);
}

const data = JSON.parse(process.argv[2]);
renderFrame(data).catch(e => { console.error(e); process.exit(1); });
